src/components/SlowPageAlert/index.tsx

```diff
@@ -15,34 +12,42 @@ export type PublicProps = {
   allowSlowPagesByDefault?: boolean;
   getLinkText: GetTextForSlowState;
   getMessage: GetTextForSlowState;
-  location: Location;
 };
 
-const SlowPageAlertBase = ({
+export type Props = PublicProps & RouteComponentProps;
+
+export const SlowPageAlertBase = ({
   _shouldAllowSlowPages = shouldAllowSlowPages,
   allowSlowPagesByDefault,
   getLinkText,
   getMessage,
+  history,
   location,
-}: PublicProps) => {
+}: Props) => {
   const allowSlowPages = _shouldAllowSlowPages({
     allowByDefault: allowSlowPagesByDefault,
     location,
   });
 
-  const newLocation = `${location.pathname}${createAdjustedQueryString(
-    location,
-    {
-      [allowSlowPagesParam]: !allowSlowPages,
-    },
-  )}`;
+  const onClick = () => {
+    const newLocation = {
+      ...location,
+      search: queryString.stringify({
+        ...queryString.parse(location.search),
+        [allowSlowPagesParam]: !allowSlowPages,
+      }),
+    };
+    history.push(newLocation);
+  };
 
   return (
     <Alert variant="danger">
       {getMessage(allowSlowPages)}{' '}
-      <Alert.Link href={newLocation}>{getLinkText(allowSlowPages)}</Alert.Link>
+      <Alert.Link onClick={onClick}>{getLinkText(allowSlowPages)}</Alert.Link>
     </Alert>
   );
 };
 
-export default SlowPageAlertBase;
+export default withRouter(SlowPageAlertBase) as React.ComponentType<
+  PublicProps
```

Comment:
Do we still have to do that? (specifying a type alias with `as`)

Reply:
I thought we needed it to tell TS only to accept `PublicProps` for the component, because we declare `Props` above to also include `RouteComponentProps`, but I see now that there's some magic in `withRouter` that takes care of that for us, so you are correct, we don't need the type alias anymore.